### PR TITLE
Encode block number as hex

### LINKED ISSUE #1270: JSON serialisation of numbers (specifically u64)
Currently we serialise `u64` (e.g. `BlockNumber`) into a plain number in the JSON over RPC, e.g. for a header we would return -

```
  "result": {
    "digest": {
      "logs": [
        {
          "Seal": [
            256804359,
            "0x2c0e0ee8a5d1a073da3f8db002b8eaebcd7b1a7eb25662e09f9ef0ac58d96bf17896256cd5f0bc672f96aab49fe5163ca5cb4c1a0f047dc39e89cd4a9eb2ea05"
          ]
        }
      ]
    },
    "extrinsicsRoot": "0x1da0d30001d07991baef3c5c65234c2e88bec659dc0d7138a7e7267e2bacbeb7",
    "number": 2918,
    "parentHash": "0xb58e4d84c2e0db1cdc8c8830aa719015f8c855f29a0dff2b158a160f29f73ba0",
    "stateRoot": "0x22f6f5b6992ab87ea5f52b2732ed9e22a88643ed3a97251986857633687fbc54"
  },
  "id": 1
}
```

(This is a good example since `BlockNumber` = `u64` and `Seal` = `(u64, Signature)`) 

This is problematic for some middleware, e.g. in JavaScript where the maximum integer number is 2^53 - 1, anything over this and the JSON will not be parsed. We have not run into this yet, since we are reasonably far from limits, however there is a possibility of problems down the line. I would suggest that anything over `u32` be encoded as the hex value.

### PR DESCRIPTION
**Pull request by tomusdrw**
Adds a custom serializer for `generic::Header` to encode block numbers as hex.
We have only one RPC that takes the block number and for consistency I decided to take hex there as well, but to also support the exact type that is declared (in case someone doesn't use `generic::Header` but has it's own type instead).
Closes #1270

### CODE REVIEW

File: core/sr-primitives/src/generic/header.rs [hunk 1]
```diff
@@ -26,10 +26,11 @@ use generic::Digest;
 #[cfg_attr(feature = "std", derive(Debug, Serialize))]
 #[cfg_attr(feature = "std", serde(rename_all = "camelCase"))]
 #[cfg_attr(feature = "std", serde(deny_unknown_fields))]
-pub struct Header<Number, Hash: HashT, DigestItem> {
+pub struct Header<Number: Copy + Into<u128>, Hash: HashT, DigestItem> {
```
Comment (rphmeier) [inline core/sr-primitives/src/generic/header.rs]: While it's unlikely that chains could have block number >= 2^128, that might be assuming that they all have a genesis with block 0 in the future.

I could see someone doing a clever "tree of chains" thing where some initial bits are used to encode a chain ID or similar.

**Comment by tomusdrw** [comment]
Just realised that the hex will be BE and the rest of substrate actually uses LE encoding more, should we use the same here? CC @gavofyork ?

**Comment by jacogr** [comment]
My comments on the BE/LE in the JSON - for the rest of the encoding, parity-codec is primarily a machine-readable representation and it certainly has benefits for the encoding/decoding (as we have seen). 

In the JSON I would be ok with BE since it is easily interpreted by machines and humans, i.e. when I do a curl to get the head and get a field containing `0x0004` back, it is human readable as 4 as well. (Have no strong preference either way around for a use perspective, just as long as we are explicit in either way - the middleware can handle whichever way we decide)

**Comment by gavofyork** [comment]
i would say BE is fine here; as jaco says, it's really meant to be a human readable alternative for numbers otherwise too large for JSON to display in the primary human-readable form.

File: core/rpc/src/chain/number.rs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+/// or we attempt to parse given hex value.
+/// We do that for consistency with the returned type, default generic header
+/// serializes block number as hex to avoid overflows in JavaScript.
+#[derive(Deserialize)]
+#[serde(untagged)]
+pub enum NumberOrHex<Number> {
+	/// The original header number type of block.
+	Number(Number),
+	/// Hex representation of the block number.
+	Hex(U256),
```
Comment (gavofyork) [inline core/rpc/src/chain/number.rs]: Bit weird having `U256` here and `u128` elsewhere...

File: core/sr-primitives/src/generic/header.rs (hunk 1 above)
Comment (gavofyork) [reply core/sr-primitives/src/generic/header.rs]: i think that as long as we require `As<u64>`, then we're good here. If a chain relied on having non-parent relationships on the same continuum of blocknumbers, then they're asking for trouble.

File: core/rpc/src/chain/number.rs (hunk 2 above)
Comment (tomusdrw) [reply core/rpc/src/chain/number.rs]: Yeah, it's actually even worse, cause we use `u64` part of that `U256` here. I'm going to submit a PR to expose `U128` from `primitive-types` crate and then unify this to have an upper bound of `u128` everywhere in the code as part of #1377

**Pull request merged**